File: js/src/xml.js
// Minimal non-validating XML parser for OOXML parts. Namespace prefixes are
// stripped from element names; attributes keep their original keys with a
// local-name lookup helper. Good for well-formed machine-written XML only.

const ENTITIES = { amp: '&', lt: '<', gt: '>', quot: '"', apos: "'" };

export function decodeEntities(s) {
  return s.replace(/&(#x?[0-9a-fA-F]+|[a-z]+);/g, (m, body) => {
    if (body[0] === '#') {
      const code = body[1] === 'x' || body[1] === 'X'
        ? parseInt(body.slice(2), 16) : parseInt(body.slice(1), 10);
      return Number.isFinite(code) ? String.fromCodePoint(code) : m;
    }
    return ENTITIES[body] ?? m;
  });
}

const TOKEN_RE = /<!\[CDATA\[([\s\S]*?)\]\]>|<!--[\s\S]*?-->|<\?[\s\S]*?\?>|<!DOCTYPE[^>]*>|<\/([^>\s]+)\s*>|<([^>\s/]+)([^>]*?)(\/?)>|([^<]+)/g;
const ATTR_RE = /([^\s=]+)\s*=\s*(?:"([^"]*)"|'([^']*)')/g;

export function parseXml(src) {
  const root = { name: '#root', attrs: {}, children: [], text: '' };
  const stack = [root];
  let m;
  TOKEN_RE.lastIndex = 0;
  while ((m = TOKEN_RE.exec(src))) {
    const top = stack[stack.length - 1];
    if (m[1] !== undefined) { top.text += m[1]; continue; }           // CDATA
    if (m[2] !== undefined) {                                          // close
      if (stack.length > 1) stack.pop();
      continue;
    }
    if (m[3] !== undefined) {                                          // open
      const el = { name: local(m[3]), attrs: {}, children: [], text: '' };
      let am;
      ATTR_RE.lastIndex = 0;
      while ((am = ATTR_RE.exec(m[4] ?? ''))) {
        el.attrs[am[1]] = decodeEntities(am[2] ?? am[3] ?? '');
      }
      top.children.push(el);
      if (!m[5]) stack.push(el);
      continue;
    }
    if (m[6] !== undefined) top.text += decodeEntities(m[6]);          // text
  }
  return root.children[0] ?? root;
}

const local = (name) => name.includes(':') ? name.slice(name.indexOf(':') + 1) : name;

// Attribute by local name ('id' matches both id and r:id — exact key wins).
export function attr(el, name) {
  if (el.attrs[name] !== undefined) return el.attrs[name];
  for (const [k, v] of Object.entries(el.attrs)) {
    if (local(k) === name) return v;
  }
  return undefined;
}

export const one = (el, name) => el.children.find((c) => c.name === name) ?? null;
export const all = (el, name) => el.children.filter((c) => c.name === name);

// Deep text of an element (its text + descendants').
export function textOf(el) {
  if (!el) return '';
  let out = el.text;
  for (const c of el.children) out += textOf(c);
  return out;
}

// First descendant (or self) with the given local name.
export function findDeep(el, name) {
  if (el.name === name) return el;
  for (const c of el.children) {
    const found = findDeep(c, name);
    if (found) return found;
  }
  return null;
}

File: rust/src/xml.rs
//! Minimal non-validating XML parser for OOXML parts. Namespace prefixes are
//! stripped from element names; attributes keep their original keys with a
//! local-name lookup helper. Port of `js/src/xml.js`. Good for well-formed
//! machine-written XML only (used by the native `from-xlsx` fallback reader).

#[derive(Debug, Clone)]
pub struct XmlNode {
    pub name: String,
    pub attrs: Vec<(String, String)>,
    pub children: Vec<XmlNode>,
    pub text: String,
}

impl XmlNode {
    fn new(name: String) -> Self {
        XmlNode {
            name,
            attrs: Vec::new(),
            children: Vec::new(),
            text: String::new(),
        }
    }

    /// Attribute by local name (`id` matches both `id` and `r:id`; exact wins).
    pub fn attr(&self, name: &str) -> Option<&str> {
        if let Some((_, v)) = self.attrs.iter().find(|(k, _)| k == name) {
            return Some(v);
        }
        self.attrs
            .iter()
            .find(|(k, _)| local(k) == name)
            .map(|(_, v)| v.as_str())
    }

    pub fn one(&self, name: &str) -> Option<&XmlNode> {
        self.children.iter().find(|c| c.name == name)
    }

    pub fn all(&self, name: &str) -> Vec<&XmlNode> {
        self.children.iter().filter(|c| c.name == name).collect()
    }

    /// Deep text of an element (its text + descendants').
    pub fn text_of(&self) -> String {
        let mut out = self.text.clone();
        for c in &self.children {
            out.push_str(&c.text_of());
        }
        out
    }
}

fn local(name: &str) -> &str {
    match name.find(':') {
        Some(i) => &name[i + 1..],
        None => name,
    }
}

pub fn decode_entities(s: &str) -> String {
    let mut out = String::new();
    let bytes = s.as_bytes();
    let mut i = 0;
    while i < bytes.len() {
        if bytes[i] == b'&' {
            if let Some(semi) = s[i..].find(';') {
                let body = &s[i + 1..i + semi];
                if let Some(ch) = decode_entity_body(body) {
                    out.push(ch);
                    i += semi + 1;
                    continue;
                }
            }
        }
        // push one UTF-8 char
        let ch = s[i..].chars().next().unwrap_or('\u{fffd}');
        out.push(ch);
        i += ch.len_utf8();
    }
    out
}

fn decode_entity_body(body: &str) -> Option<char> {
    if let Some(rest) = body.strip_prefix('#') {
        let code = if let Some(hex) = rest.strip_prefix(['x', 'X']) {
            u32::from_str_radix(hex, 16).ok()?
        } else {
            rest.parse::<u32>().ok()?
        };
        return char::from_u32(code);
    }
    match body {
        "amp" => Some('&'),
        "lt" => Some('<'),
        "gt" => Some('>'),
        "quot" => Some('"'),
        "apos" => Some('\''),
        _ => None,
    }
}

/// Parse XML into a tree. Returns the first top-level element (mirrors xml.js).
pub fn parse_xml(src: &str) -> XmlNode {
    let mut root = XmlNode::new("#root".to_string());
    // Stack of node paths (indices from root).
    let mut stack: Vec<XmlNode> = vec![root];
    let chars: Vec<char> = src.chars().collect();
    let mut i = 0;
    let n = chars.len();
    while i < n {
        if chars[i] == '<' {
            // Special constructs.
            if starts_with(&chars, i, "<![CDATA[") {
                let end = find_seq(&chars, i + 9, "]]>");
                let content: String = chars[i + 9..end].iter().collect();
                stack.last_mut().unwrap().text.push_str(&content);
                i = end + 3;
                continue;
            }
            if starts_with(&chars, i, "<!--") {
                let end = find_seq(&chars, i + 4, "-->");
                i = end + 3;
                continue;
            }
            if starts_with(&chars, i, "<?") {
                let end = find_seq(&chars, i + 2, "?>");
                i = end + 2;
                continue;
            }
            if starts_with(&chars, i, "<!") {
                let end = find_char(&chars, i, '>');
                i = end + 1;
                continue;
            }
            if starts_with(&chars, i, "</") {
                let end = find_char(&chars, i, '>');
                if stack.len() > 1 {
                    let node = stack.pop().unwrap();
                    stack.last_mut().unwrap().children.push(node);
                }
                i = end + 1;
                continue;
            }
            // Open tag.
            let end = find_char(&chars, i, '>');
            let inner: String = chars[i + 1..end].iter().collect();
            let self_close = inner.ends_with('/');
            let inner = inner.strip_suffix('/').unwrap_or(&inner);
            let (name, attrs) = parse_tag(inner);
            let mut el = XmlNode::new(local(&name).to_string());
            el.attrs = attrs;
            if self_close {
                stack.last_mut().unwrap().children.push(el);
            } else {
                stack.push(el);
            }
            i = end + 1;
        } else {
            let start = i;
            while i < n && chars[i] != '<' {
                i += 1;
            }
            let text: String = chars[start..i].iter().collect();
            let decoded = decode_entities(&text);
            stack.last_mut().unwrap().text.push_str(&decoded);
        }
    }
    // Unwind any unclosed nodes.
    while stack.len() > 1 {
        let node = stack.pop().unwrap();
        stack.last_mut().unwrap().children.push(node);
    }
    root = stack.pop().unwrap();
    root.children.into_iter().next().unwrap_or_else(|| XmlNode::new("#root".to_string()))
}

fn parse_tag(inner: &str) -> (String, Vec<(String, String)>) {
    let inner = inner.trim();
    // name is up to first whitespace
    let (name, rest) = match inner.find(|c: char| c.is_whitespace()) {
        Some(p) => (&inner[..p], &inner[p..]),
        None => (inner, ""),
    };
    let mut attrs = Vec::new();
    let bytes: Vec<char> = rest.chars().collect();
    let mut i = 0;
    while i < bytes.len() {
        while i < bytes.len() && bytes[i].is_whitespace() {
            i += 1;
        }
        if i >= bytes.len() {
            break;
        }
        let key_start = i;
        while i < bytes.len() && bytes[i] != '=' && !bytes[i].is_whitespace() {
            i += 1;
        }
        let key: String = bytes[key_start..i].iter().collect();
        while i < bytes.len() && bytes[i].is_whitespace() {
            i += 1;
        }
        if i < bytes.len() && bytes[i] == '=' {
            i += 1;
            while i < bytes.len() && bytes[i].is_whitespace() {
                i += 1;
            }
            if i < bytes.len() && (bytes[i] == '"' || bytes[i] == '\'') {
                let q = bytes[i];
                i += 1;
                let val_start = i;
                while i < bytes.len() && bytes[i] != q {
                    i += 1;
                }
                let val: String = bytes[val_start..i].iter().collect();
                i += 1; // closing quote
                if !key.is_empty() {
                    attrs.push((key, decode_entities(&val)));
                }
            }
        } else if !key.is_empty() {
            // valueless attribute — ignore
        }
    }
    (name.to_string(), attrs)
}

fn starts_with(chars: &[char], i: usize, pat: &str) -> bool {
    let p: Vec<char> = pat.chars().collect();
    if i + p.len() > chars.len() {
        return false;
    }
    chars[i..i + p.len()] == p[..]
}

fn find_seq(chars: &[char], from: usize, pat: &str) -> usize {
    let p: Vec<char> = pat.chars().collect();
    let mut i = from;
    while i + p.len() <= chars.len() {
        if chars[i..i + p.len()] == p[..] {
            return i;
        }
        i += 1;
    }
    chars.len()
}

fn find_char(chars: &[char], from: usize, c: char) -> usize {
    let mut i = from;
    while i < chars.len() && chars[i] != c {
        i += 1;
    }
    i
}
